lib/Parse/ParseExpr.cpp

```diff
@@ -2995,3 +3011,63 @@ Parser::parseVersionConstraintSpec() {
   return makeParserResult(new (Context) VersionConstraintAvailabilitySpec(
       Platform.getValue(), PlatformLoc, Version, VersionRange));
 }
+
+/// parseExprTypeOf
+///
+///   expr-dynamictype:
+///     'type' '(' 'of:' expr ')'
+///
+ParserResult<Expr> Parser::parseExprTypeOf() {
+
+  assert(Tok.getText() == "type" && "only 'type' should be handled here");
+
+  // Consume 'type'
+  SourceLoc keywordLoc = consumeToken();
+  // Consume '('
+  SourceLoc lParenLoc = consumeToken(tok::l_paren);
+
+  // Parse `of` label
+  // If we see a potential argument label followed by a ':', consume
+  // it.
+  if (Tok.canBeArgumentLabel() && Tok.getText() == "of" && peekToken().is(tok::colon)) {
+    consumeToken();
+    consumeToken(tok::colon);
+  } else {
+    diagnose(Tok, diag::expr_typeof_expected_label_of);
+    return makeParserError();
+  }
+
+  // Parse the subexpression.
+  ParserResult<Expr> subExpr = parseExpr(diag::expr_typeof_expected_expr);
+  if (subExpr.hasCodeCompletion())
+    return makeParserCodeCompletionResult<Expr>();
+
+  // Parse the closing ')'
+  SourceLoc rParenLoc;
+  if (subExpr.isParseError()) {
+    skipUntilDeclStmtRBrace(tok::r_paren);
+    if (Tok.is(tok::r_paren))
+      rParenLoc = consumeToken();
+    else
+      rParenLoc = Tok.getLoc();
+  } else {
+    parseMatchingToken(tok::r_paren, rParenLoc,
+                       diag::expr_typeof_expected_rparen, lParenLoc);
+  }
+
+  // If the subexpression was in error, just propagate the error.
+  if (subExpr.isParseError()) {
+    if (subExpr.hasCodeCompletion()) {
+      auto res = makeParserResult(
+                                  new (Context) DynamicTypeExpr(subExpr.get(), consumeToken(), Type()));
```

Comment:
Outdent this to be aligned two spaces after `makeParserResult` and break it across multiple lines.